src/services/events/index.ts

```diff
@@ -0,0 +1,123 @@
+import { CartDto, Product, User } from '../../interfaces'
+
+type EventCallbackFunction<Payload = any> = (payload: Payload) => void
+
+type GetTypeFromInterface<T> = T extends EventCallbackFunction<infer Payload> ? Payload : undefined
+
+type EventsListenerService = () => {
+  emit: <Key extends keyof Events>(
+    event: Key,
+    payload?: GetTypeFromInterface<Events[Key][0]> | undefined,
+  ) => void
+  on: (event: keyof Events, cb: EventCallbackFunction) => void
+}
+
+export interface Events {
+  /**
+   * The addition of an item to a shopping cart or basket.
+   */
+  addToCart: EventCallbackFunction<Product>[]
```

Comment:
Klucze eventów fajny byłoby wyciągnąć do enuma, łatwiej będzie się to potem obsługiwać

Reply:
Nie wyciągnąłeś do enumów